src/core/provider.h

```diff
@@ -35,6 +35,42 @@ namespace nvidia { namespace inferenceserver {
 
 class InferenceBackend;
 
+//
+// Memory block used for accessing / storing data for providers
+// Note:
+//   - It is only responsible for deallocating memory obtained by ReserveBlock()
+//   - The functions are not thread-safe
+//
+class SystemMemoryBlock {
+ public:
+  // Create an SystemMemoryBlock object with block size 0
+  SystemMemoryBlock();
+
+  // Create an SystemMemoryBlock that points to 'block' provided
+  // 'block' is the memory block that has been allocated. For instance,
+  // content returned by InferRequestProvider.GetNextInputContent().
+  // 'block_size' is the size of the block
+  SystemMemoryBlock(void* block, size_t block_size);
+
+  // Reserve a block of memory with size 'byte_size'
+  // 'byte_size' is the size of the memory to be reserved
+  // Return error if the SystemMemoryBlock has reserved memory
+  // with non-zero size
+  Status ReserveBlock(size_t byte_size);
+
+  // Return a pointer to the memory location. It is caller's responsibility
+  // to make sure the access to the memory is exclusive on write.
+  void* GetBlock() const;
+
+  // Return the byte size of the memory
+  size_t ByteSize() const { return byte_size_; }
+
+ private:
+  size_t byte_size_;
+  std::unique_ptr<char[]> buffer_;
+  void* referencing_block_;
```

Comment:
This class seems more complicated than it needs to be. Why is it not just a size and buffer? A response provider will create it and wrap it in a shared_ptr. Then when the sceduler is creating the request providers that will provider this output to the "next" models, it simply makes a copy of the shared_ptr. I don't see what referencing_block_ and ReserveBlock are needed for.

Reply:
I see. referencing_block_ is for wrapping the data in the HTTP/GRPC request provider, so we don't need to copy them to create the provider for actual models.

But yes, ReserveBlock will not be needed if the response provide is creating the object.